protocols/relay/src/v2/client/handler.rs

```diff
@@ -0,0 +1,610 @@
+            return Poll::Ready(event);
+        }
+
+        // Check if reservation needs renewal.
+        match self.reservation.take() {
+            Some(Reservation::Accepted {
+                mut renewal_timeout,
+                pending_msgs,
+                to_listener,
+            }) => match renewal_timeout.as_mut().map(|t| t.poll_unpin(cx)) {
+                Some(Poll::Ready(())) => {
+                    self.reservation = Some(Reservation::Renewal { pending_msgs });
+                    return Poll::Ready(ProtocolsHandlerEvent::OutboundSubstreamRequest {
+                        protocol: SubstreamProtocol::new(
+                            outbound_hop::Upgrade::Reserve,
+                            OutboundOpenInfo::Reserve { to_listener },
+                        ),
+                    });
+                }
+                Some(Poll::Pending) | None => {
+                    self.reservation = Some(Reservation::Accepted {
+                        renewal_timeout,
+                        pending_msgs,
+                        to_listener,
+                    });
+                }
+            },
+            r => self.reservation = r,
+        }
+
+        // Forward messages to transport listener.
+        if let Some(Reservation::Accepted {
+            pending_msgs,
+            to_listener,
+            ..
+        }) = &mut self.reservation
+        {
+            if !pending_msgs.is_empty() {
+                match to_listener.poll_ready(cx) {
+                    Poll::Ready(Ok(())) => {
+                        if let Err(e) = to_listener
+                            .start_send(pending_msgs.pop_front().expect("Called !is_empty()."))
+                        {
+                            debug!("Failed to sent pending message to listener: {:?}", e);
+                            self.reservation.take();
+                        }
+                    }
+                    Poll::Ready(Err(e)) => {
+                        debug!("Channel to listener failed: {:?}", e);
+                        self.reservation.take();
+                    }
+                    Poll::Pending => {}
+                }
+            }
+        }
+
+        // Deny incoming circuit requests.
+        if let Poll::Ready(Some((src_peer_id, result))) = self.circuit_deny_futs.poll_next_unpin(cx)
+        {
+            match result {
+                Ok(()) => {
+                    return Poll::Ready(ProtocolsHandlerEvent::Custom(
+                        Event::InboundCircuitReqDenied { src_peer_id },
+                    ))
+                }
+                Err(error) => {
+                    return Poll::Ready(ProtocolsHandlerEvent::Custom(
+                        Event::InboundCircuitReqDenyFailed { src_peer_id, error },
+                    ))
+                }
+            }
+        }
+
+        // Send errors to transport.
+        while let Poll::Ready(Some(())) = self.send_error_futs.poll_next_unpin(cx) {}
```

Comment:
Sorry, I'm technically wrong. This will get called sometimes but only when `self.pending_error` is `None`.

What makes this tricky/surprising is that the code [here](https://github.com/libp2p/rust-libp2p/pull/2059/files#diff-f237102af98d25da24bc739a72d2f557a689d8cb1256485b1a587570bd4fab1bR395) makes it seem like we expect to send the error to the the transport. But in reality there are only a [couple](https://github.com/libp2p/rust-libp2p/pull/2059/files#diff-f237102af98d25da24bc739a72d2f557a689d8cb1256485b1a587570bd4fab1bR388) of [cases](https://github.com/libp2p/rust-libp2p/pull/2059/files#diff-f237102af98d25da24bc739a72d2f557a689d8cb1256485b1a587570bd4fab1bR336-R339) where we would put something in `self.send_error_futs` without also setting `self.pending_error` and thus actually advancing `self.send_error_futs`. This is related to our discussion here: https://github.com/libp2p/rust-libp2p/pull/2059/files#r763386312.

I'm not sure what the correct behavior should be, but I would expect we want to send the transport the messages we queue up in the error path. So maybe we should move this to happen before we return the `self.pending_error`? Although on the other hand we don't want to prevent this handler from closing because we are still awaiting to send errors to the transport... so maybe we should be okay with dropping errors to the transport.

If we're okay dropping errors, and if we really only do want to send errors to the transport when we don't have a `self.pending_error` then I think we should wrap [this](https://github.com/libp2p/rust-libp2p/pull/2059/files#diff-f237102af98d25da24bc739a72d2f557a689d8cb1256485b1a587570bd4fab1bR395) with a conditional (i.e. `if self.pending_error.is_none() { self.send_error_futs.push(...) }`) to make it more obvious that we only expect that to advance if there is no pending_error. And maybe add a comment [here](https://github.com/libp2p/rust-libp2p/pull/2059/files#diff-f237102af98d25da24bc739a72d2f557a689d8cb1256485b1a587570bd4fab1bR161) explaining the behavior. i.e. `// Futures that try to send errors to the transport. We may drop errors if this handler ends up in a terminal state (by returning ProtocolsHandlerEvent)`

Does that make sense? Maybe this was obvious, but I found this behavior surprising/tricky. also open to other ideas here.

Reply:
Thanks for elaborating! I understand the confusion. 746c75ee40b525e91896c436cdf183fada3a1ec1 applies your suggestions. It is not ideal, but hopefully now easier for newcomers to understand.